### PR TITLE
Add cache warmup endpoint

### PR DESCRIPTION
**Pull request by bkyryliuk**
Implements the FR: https://github.com/airbnb/caravel/issues/1021
Based on the PR: https://github.com/airbnb/caravel/pull/1062

This PR adds 2 types of the cache warm up endpoints:
- Via table name and db name: 
  /caravel/warm_up_cache?table_name=energy_usage&db_name=main
- Via slice id: /caravel/warm_up_cache?slice_id=1

### CODE REVIEW

File: caravel/models.py [hunk 1]
```diff
@@ -250,6 +252,34 @@ def slice_link(self):
         return '<a href="{url}">{obj.slice_name}</a>'.format(
             url=url, obj=self)
 
+    def get_viz(self, url_params_multidict=None):
+        """Creates BaseViz object from the url_params_multidict.
+
+        Parameters
```
Comment (mistercrunch) [inline caravel/models.py]: If we want to auto generate docs for the code we should stick with Sphinx's (the documentation framework we use) way of specifying docstrings:
http://www.sphinx-doc.org/en/stable/domains.html#info-field-lists

File: caravel/views.py [hunk 2]
```diff
@@ -901,130 +941,89 @@ def explore(self, datasource_type, datasource_id, slice_id=None):
...
-            # override slice params with anything passed in url
-            # some overwritten slices have been saved with original slice_ids
-            param_dict["slice_id"] = slice_id
-            param_dict["json"] = "false"
-            param_dict.update(request.args.to_dict(flat=False))
-            slice_params = ImmutableMultiDict(param_dict)
+                flash(utils.error_msg_from_exception(e), "danger")
+                return redirect(error_redirect)
+        else:
+            viz_type = request_args_multi_dict.get("viz_type", None)
```
Comment (mistercrunch) [inline caravel/views.py]: Unless multidict has a subtlety when compared to a normal dict, the second argument `None` is implicit as the get method returns None if the key is not found by default

File: caravel/views.py [hunk 3]
```diff
@@ -901,130 +941,89 @@ def explore(self, datasource_type, datasource_id, slice_id=None):
...
+            if not viz_type and datasource.default_endpoint:
+                return redirect(datasource.default_endpoint)
+            # default to table if no default endpoint and no viz_type
+            viz_type = viz_type if viz_type else "table"
```
Comment (mistercrunch) [inline caravel/views.py]: pythonesquier:
`viz_type = viz_type or 'table'`

File: caravel/views.py [hunk 4]
```diff
@@ -901,130 +941,89 @@ def explore(self, datasource_type, datasource_id, slice_id=None):
...
-
-            resp = self.render_template(
-                template, viz=obj, slice=slc, datasources=datasources,
+            return self.render_template(
+                template, viz=viz_obj, slice=slc, datasources=datasources,
                 can_add=slice_add_perm, can_edit=slice_edit_perm,
                 can_download=slice_download_perm,
                 userid=g.user.get_id() if g.user else '')
 
-            try:
```
Comment (mistercrunch) [inline caravel/views.py]: oh wow, I'm not sure why we had that weird empty block here. Looks like a bad merge of some sort...

File: caravel/views.py [hunk 5]
```diff
@@ -1238,6 +1237,48 @@ def testconn(self):
                 status=500,
                 mimetype="application/json")
 
+    @api
+    @has_access_api
+    @expose("/warm_up_cache/", methods=['GET', 'POST'])
+    def warm_up_cache(self):
+        """Warms up the cache for the slice or table."""
+        slices = None
+        s = db.session()
```
Comment (mistercrunch) [inline caravel/views.py]: `s` is commonly used for short lived strings, I'd go with `session` instead to avoid confusion

File: caravel/views.py [hunk 6]
```diff
@@ -1238,6 +1237,48 @@ def testconn(self):
...
+            return json_error_response(__(
+                "Malformed request. slice_id or table_name and db_name "
+                "arguments are expected"), status=400)
+        if slice_id:
+            slices = s.query(models.Slice).filter_by(id=slice_id).all()
+            if not slices:
+                return json_error_response(__(
+                    "Slice %(id)s not found", id=slice_id), status=404)
+        elif table_name and db_name:
+            table = s.query(models.SqlaTable).join(models.Database).filter(
```
Comment (mistercrunch) [inline caravel/views.py]: NIT: the way I've been going about breaking sqla queries is to wrap the statement in parenthesis, and then break before the dot as in javascript when doing method call chains, makes it more readable. Example:
```
foo = (
    session.query(BarModel)
    .filterby(
        id==id,
        bar_style='foobar',
    )
    .all()
```

File: caravel/views.py [hunk 7]
```diff
@@ -1238,6 +1237,48 @@ def testconn(self):
                 status=500,
                 mimetype="application/json")
 
+    @api
+    @has_access_api
+    @expose("/warm_up_cache/", methods=['GET', 'POST'])
```
Comment (mistercrunch) [inline caravel/views.py]: Since you read from `request.args`, I think it will only work with `GET`

**Comment by mistercrunch** [comment]
Nice. A few minor nits but the code looks much better overall!

**Comment by mistercrunch** [comment]
Oh looks like this PR is a superset of https://github.com/airbnb/caravel/pull/1062 isn't it? I won't review 1062 if that's the case.

File: caravel/views.py (hunk 3 above)
Comment (bkyryliuk) [inline caravel/views.py]: that's magical ) didn't know about this python feature

File: caravel/views.py (hunk 7 above)
Comment (bkyryliuk) [inline caravel/views.py]: you're right, changed the method to be GET

File: caravel/views.py (hunk 6 above)
Comment (bkyryliuk) [inline caravel/views.py]: sounds good, I'll structure my queries this way

**Comment by bkyryliuk** [comment]
@mistercrunch - yes. I wasn't sure how stack the PRs on top of each other here.

**Comment by mistercrunch** [comment]
LGTM!

**Comment by bkyryliuk** [comment]
thanks @mistercrunch for the review!

**Pull request merged**